src/sknnr/transformers/_cca_transformer.py

```diff
@@ -8,30 +8,25 @@
 
 
 class CCATransformer(ComponentReducerMixin, TransformerMixin, BaseEstimator):
-    @property
-    def _n_features_out(self):
-        return self.n_components_
-
     def get_feature_names_out(self) -> NDArray:
```

Comment:
Brilliant!  I like this idea, but could use your help.  I've taken `get_feature_names_out` of both `CCATransformer` and `CCorATransformer` and put it into `ComponentReducerMixin` like this:
```
    def get_feature_names_out(self) -> NDArray:
        check_is_fitted(self, ["ordination_", "n_components_"])
        return np.asarray(
            [
                f"{self.ordination_.__class__.__name__.lower()}{i}"
                for i in range(self.n_components_)
            ],
            dtype=object,
        )
```
One issue, two questions:
1. mypy isn't passing pre-commit based on this message: `error: "ComponentReducerMixin" has no attribute "ordination_"  [attr-defined]`.  Because `ordination_` is getting set in the `fit` method of the transformers, `ComponentReducerMixin` doesn't see it.  I feel like you've solved this issue before - do I have to declare it somewhere in `ComponentReducerMixin`?
2. Do you want a custom error message on `check_is_fitted` or will the default one work: `“This %(name)s instance is not fitted yet. Call ‘fit’ with appropriate arguments before using this estimator.”`?
3. How do you show co-authorship on a commit?  I've seen that you've done this, but I wasn't sure how to do this.  I'm getting quite a bit of help here from you on this PR!

Reply:
>  I feel like you've solved this issue before - do I have to declare it somewhere in ComponentReducerMixin?

Yep, here's that line in `TransformedKNeighborsMixin`: https://github.com/lemma-osu/scikit-learn-knn-regression/blob/55e079d913f8deb939b57bdb8cad3bedcae14c17/src/sknnr/_base.py#L44

Since the ordinations don't share a superclass, I guess the declaration should be `ordination_: CCorA | CCA`? Or that could be declared as a module-level type, e.g. `Ordination = CCorA | CCA`. I could go either way.

This is probably too big of an issue to tackle here, but I wonder if those could/should be refactored to share an `Ordination` superclass? I'm not sure if they really share enough functionality for that to make sense.

> Do you want a custom error message on check_is_fitted or will the default one work: “This %(name)s instance is not fitted yet. Call ‘fit’ with appropriate arguments before using this estimator.”?

The default's fine with me - that's what `PCA.get_feature_names_out` does.

> How do you show co-authorship on a commit? I've seen that you've done this, but I wasn't sure how to do this. I'm getting quite a bit of help here from you on this PR!

Don't worry about it! I think that only happens when someone suggests a specific code change that gets committed :)

Reply:
> This is probably too big of an issue to tackle here, but I wonder if those could/should be refactored to share an `Ordination` superclass? I'm not sure if they really share enough functionality for that to make sense.

Oh, absolutely!  And there are a couple more ordination techniques that are even more closely related to CCA that would make this a no-brainer.  Famous last words, but I fully intend to refactor those two classes and find some commonalities between them and having an `Ordination` base class for these makes a lot of sense.  Not for this PR, but I will start an issue.

> Since the ordinations don't share a superclass, I guess the declaration should be `ordination_: CCorA | CCA`? Or that could be declared as a module-level type, e.g. `Ordination = CCorA | CCA`. I could go either way.

I went with the latter with the idea that if we *do* create a base class, we'll just remove this line and import it from elsewhere.  Two other comments:

1. You can probably see that I had a failing test, because I tried to use `CCorA | CCA` syntax rather than `Union[CCorA, CCA]` syntax.  I tried with `from __future__ import annotations`, but that doesn't cover the union operator, correct?
2. Check me on this, but because we now have `ordination_` as a class attribute, it makes no sense to pass it to `check_is_fitted`, because it is only checking for the presence of that estimator attribute, correct?

Reply:
> I went with the latter with the idea that if we do create a base class, we'll just remove this line and import it from elsewhere.

Good thinking!

>  I tried with from __future__ import annotations, but that doesn't cover the union operator, correct?

Interesting, the tests failed with that import in place? I just switch from `Union` to `|` for my type annotations in [sankee](https://github.com/aazuspan/sankee/blob/1242246d25b8e15918bea8faec735270c316dd93/sankee/sampling.py#L16) and the `3.8` tests pass there as long as I have the `__future__` import. It would be nice to have `tox` set up to be able to reproduce these version-specific bugs locally...

> Check me on this, but because we now have ordination_ as a class attribute, it makes no sense to pass it to check_is_fitted, because it is only checking for the presence of that estimator attribute, correct?

Yeah, good point. Unless there's some hypothetical case where it could be have some fitted attributes but not others (not a possibility here), I don't think there's any particular reason to specify those parameters.

Reply:
> Interesting, the tests failed with that import in place? I just switch from `Union` to `|` for my type annotations in [sankee](https://github.com/aazuspan/sankee/blob/1242246d25b8e15918bea8faec735270c316dd93/sankee/sampling.py#L16) and the `3.8` tests pass there as long as I have the `__future__` import. It would be nice to have `tox` set up to be able to reproduce these version-specific bugs locally...

I hadn't explicitly tried it before, but I just committed and pushed with the union operator in there and it [does bonk](https://github.com/lemma-osu/scikit-learn-knn-regression/actions/runs/5417534136/jobs/9848606292?pr=40).  Maybe I'm doing something wrong?  (I went ahead and reverted to `Union[CCA, CCorA]`)

> Unless there's some hypothetical case where it could be have some fitted attributes but not others (not a possibility here), I don't think there's any particular reason to specify those parameters

I want to make sure I understand.  Currently, I'm still checking for `n_components_`, e.g.
```
check_is_fitted(self, "n_components_")
```
but are you saying we should just not specify the attributes parameter, in which case it will just verify that a single estimator attribute is present?

Reply:
> I hadn't explicitly tried it before, but I just committed and pushed with the union operator in there and it [does bonk](https://github.com/lemma-osu/scikit-learn-knn-regression/actions/runs/5417534136/jobs/9848606292?pr=40). Maybe I'm doing something wrong? (I went ahead and reverted to Union[CCA, CCorA])

No, you were right! Thanks for checking. Apparently there's a difference in behavior (at least in 3.8) between type annotations within a function (like I used in sankee) and type aliases (like we're using here), we I think we're stuck with `Union`. Here's a relevant [snippet from the mypy docs](https://mypy.readthedocs.io/en/stable/runtime_troubles.html#future-annotations-import-pep-563) that vaguely mentions this:

![image](https://github.com/lemma-osu/scikit-learn-knn-regression/assets/50475791/62573850-76f0-429f-9868-d431f2ac521e)

As you said before, typing is hard!

> but are you saying we should just not specify the attributes parameter, in which case it will just verify that a single estimator attribute is present?

That was what I was saying, but your solution of checking for `n_components_` is more explicit and is consistent with the [check in PCA](https://github.com/scikit-learn/scikit-learn/blob/364c77e047ca08a95862becf40a04fe9d4cd2c98/sklearn/base.py#L944), so your instinct was right!